### PR TITLE
Swappable parsing

### PR DESCRIPTION
**Pull request by rmitsch**
## Description

Make parsing a swappable, configurable attribute to allow modify parsing behaviour without retiring the entire task class.

### Corresponding documentation PR

todo

### Types of change

New feature.

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran all tests in `tests` and `usage_examples/tests`, and all new and existing tests passed. This includes
  - all external tests (i. e. `pytest` ran with `--external`)
  - all tests requiring a GPU (i. e. `pytest` ran with `--gpu`)
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: spacy_llm/tasks/lemma.py
Comment (rmitsch) [inline spacy_llm/tasks/lemma.py:101]: This would be temporary. If we decide to do this, all response parse functions would be registered.

File: spacy_llm/ty.py
Comment (kabirkhan) [inline spacy_llm/ty.py:19]: this is complex enough for a protocol and I think `PromptExecutor` should probably be one as well.

Comment (rmitsch) [reply spacy_llm/ty.py:19]: Hm, I don't see the benefit of that. IIRC protocols are tricky w.r.t. typing if only a function (and not a callable object) is supplied. Independent from that I see protocols only making sense for classes expected to have at least one method with a certain signature. IMO a type definition is the simplest tool for the job here.

I'm interested in hearing your arguments in favor of using protocols however, I might be missing something.

**Comment by rmitsch** [comment]
It might make sense to move each task into their own module and split the actual task implementation from the factory methods and utils (such as the parsing methods). Over time this will become unwieldy otherwise.

**Comment by rmitsch** [comment]
Making parsing swappable is a bit more complicated for tasks inheriting from `SpanTask`, as there other class methods are invoked. There isn't a strong reason for that though, this setup works fine with moving all parsing-related functions out of the class and registering them (which is what this PR aims for, ultimately).

**Review by svlandeg — COMMENTED** [review]
This looks like a promising approach to me to allow more flexibility both for users, as well as for our own builtin functionality.

As discussed internally, we'd also need to make the `_create_prompt_example` functionality swappable.

**Comment by svlandeg** [comment]
> It might make sense to move each task into their own module and split the actual task implementation from the factory methods and utils (such as the parsing methods). Over time this will become unwieldy otherwise.

We can definitely do that, yep.

File: spacy_llm/ty.py
Comment (svlandeg) [reply spacy_llm/ty.py:19]: I agree with Raphael that a type definition seems fine here.

File: spacy_llm/tasks/lemma/registry.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/registry.py:6]: I think `util` is not a good name for this, as it's not just boilerplate utility code, this is specific functionality. Maybe a `parser.py` and `examples.py`. And read the templates in `task.py` (not married to any of these names/decisions though)

File: spacy_llm/tasks/lemma/registry.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/registry.py:16]: Are these supposed to be the same?

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:17]: I think we should remove the default template here and assume it's always defined?

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:18]: I don't think this can be `Optional`? It'd have to be defined, right?

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:29]: `fewshot_example_type` isn't documented in the docstring

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:19]: Again, not sure how this can be `Optional` if the code asserts it to be not `None`?

**Review by svlandeg — COMMENTED** [review]
I like this direction! Leaving some first comments.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:17]: We do have duplicated default values for all tasks. Duplicated as in default values are defined in the factory function _and_ the task. Should we just remove all default values from all task constructors?

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:19]: It's set in `__init__()`.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:30]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/047a496ca863c972e2a35600b6492960cac81706.

File: spacy_llm/tasks/lemma/registry.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/registry.py]: The latter is redundant, forgot to remove it. Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/047a496ca863c972e2a35600b6492960cac81706.

File: spacy_llm/tasks/lemma/registry.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/registry.py]: Done in https://github.com/explosion/spacy-llm/pull/242/commits/bd416bafc2bf4461791dc666c790502af9dc430e.

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [reply spacy_llm/tasks/lemma/task.py:20]: Urgh. Right.

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [reply spacy_llm/tasks/lemma/task.py:21]: Hm, ok. Should we move the `assert` to `__init__()` though?

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [reply spacy_llm/tasks/lemma/task.py:19]: I think so? Like in spaCy's factories we also only define the defaults in the `default_config`'s, right?

File: spacy_llm/tasks/rel.py
Comment (adrianeboyd) [inline spacy_llm/tasks/rel.py:10]: I think you want to move all these imports into the `try`/`except` block because they might not interact like you'd expect? (I am not 100% sure, but it could be the source of the warnings. I don't understand why it's only in some python versions, though.)

Comment (rmitsch) [reply spacy_llm/tasks/rel.py]: I'm also wondering why it's only on Mac with Python 3.8 :upside_down_face: Done in https://github.com/explosion/spacy-llm/pull/242/commits/3bcf2a8339439cc268cee055887c3fec3aa5b6c4.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:21]: Done in https://github.com/explosion/spacy-llm/pull/242/commits/0f869a729b46ffbfb9e23fcf57b63be1ce1adbd4.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:19]: From what I've seen (e. g. [`EntityLinker`](https://github.com/explosion/spaCy/blob/master/spacy/pipeline/entity_linker.py#L168)), yes.

File: spacy_llm/tasks/rel.py
Comment (rmitsch) [reply spacy_llm/tasks/rel.py]: Done in https://github.com/explosion/spacy-llm/pull/242/commits/19410177bd8630c55ed1f58529cf4c9091fb3a02. Seems to get rid of the warnings indeed - good call, thank you!

File: spacy_llm/tasks/lemma/registry.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/registry.py:41]: I suggest to either shorten both (as below), or spell out both before the `return` statement:

```suggestion
        parse_responses=parse_responses or parse_responses_v1,
        fewshot_example_type=fewshot_example_type or LemmaExample,
```

I wouldn't use the `or` style for more complicated things like the `lemma_examples` definition though.

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:13]: It bothers me that we didn't version the v1's in their name, can we still change that for the sake of consistency?

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:39]: Let's discuss this a bit more internally - it feels weird to type these as `Optional` and then immediately `assert` that they're not `None`. It's all related to the serialization & needing `None` defaults.

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:57]:
```suggestion
```

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:77]:
```suggestion
```
(moved to `__init__`)

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:87]: Wait why isn't this `"lemma"`?

File: spacy_llm/tasks/lemma/task.py
Comment (svlandeg) [inline spacy_llm/tasks/lemma/task.py:95]:
```suggestion
```

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:57]: Needed for `mypy` to be happy, unfortunately. Alternatively I can add a `mypy` ignore comment.

Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:57]: Actually, this should be rendered redundant anyway once this isn't optional anymore.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py:87]: Good catch :smile:

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/46fc5cfc0f498ef54a68a3910e5c0653cbcf20dd.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/46fc5cfc0f498ef54a68a3910e5c0653cbcf20dd.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/46fc5cfc0f498ef54a68a3910e5c0653cbcf20dd.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/46fc5cfc0f498ef54a68a3910e5c0653cbcf20dd.

File: spacy_llm/tasks/lemma/task.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/task.py]: Updated in https://github.com/explosion/spacy-llm/pull/242/commits/43368c3f94f11d97f2e2efa94205a201dc8d6129.

**Comment by rmitsch** [comment]
Failing tests are expected at this point, as I've started, but haven't finished, working on tasks derived from `SpanTask`.

File: spacy_llm/tasks/ner/registry.py
Comment (svlandeg) [inline spacy_llm/tasks/ner/registry.py:29]:
```suggestion
```

File: spacy_llm/tasks/ner/registry.py
Comment (svlandeg) [inline spacy_llm/tasks/ner/registry.py:33]:
```suggestion
```

**Review by svlandeg — COMMENTED** [review]
Mostly this looks good!

Let's 
- Fix the merge conflict
- Double check the docs strings for consistency
- get the refactor PR merged in

And then I can do a final review round?

File: spacy_llm/tasks/span/parser.py
Comment (kabirkhan) [inline spacy_llm/tasks/span/parser.py:35]: I don't think `kwargs` makes sense here. It actually loses quite a bit of clarity in the function. I'm assuming this is done so we have a single `TaskResponseParserProtocol`.

If you're looking at this function in isolation, we don't validate `kwargs` at all, we just assume these keys are present. This "works" but is not very easy to understand and is a source of potential bugs.

File: spacy_llm/tasks/ner/task.py
Comment (kabirkhan) [inline spacy_llm/tasks/ner/task.py:24]: This type can't be `SpanExample` because `SpanExample` assumes the `text`, and `entities` fields. The COT prompting has different types for the keys and wouldn't be compatible.

File: spacy_llm/tasks/span/parser.py
Comment (rmitsch) [reply spacy_llm/tasks/span/parser.py:35]: I agree on the merits of explicitly stating arguments over `kwargs`. I originally went for the latter because I couldn't find another way to make the types of `_parse_responses` agree with the parent's `__init__()` signature. I'll have another look into how to at least approximate a typing match there.

File: spacy_llm/tasks/ner/task.py
Comment (rmitsch) [reply spacy_llm/tasks/ner/task.py:23]: In this case `SpanTask` will have to be modified as well, as it relies on `entities` and `text` being attributes of `SpanExample`. I suggest approaching this in the CoT PR though.

File: spacy_llm/tasks/ner/task.py
Comment (rmitsch) [reply spacy_llm/tasks/ner/task.py]: Do we want to call this attribute `prompt_examples` or `fewshot_examples` or `examples`? This is a inconsistent currently.

File: spacy_llm/tasks/span/parser.py
Comment (kabirkhan) [reply spacy_llm/tasks/span/parser.py:35]: Check out `ParamSpec`, that's the general way to type a callable and say inherit these arguments

File: spacy_llm/ty.py
Comment (rmitsch) [reply spacy_llm/ty.py:92]: We have `ScorerProtocol` already. These two are used for different purposes. I'm open to renaming suggestions.

Comment (rmitsch) [reply spacy_llm/ty.py:92]: Eh, I meant we have `ScorableProtocol` already. The difference between `ScorableProtocol` and `ScorerProtocol` isn't significant enough IMO.

File: spacy_llm/tasks/span/parser.py
Comment (rmitsch) [reply spacy_llm/tasks/span/parser.py:35]: Good point, thanks. Not sure if after transitioning to explicit signatures for the concrete `parse_responses()` functions using `ParamSpec` over `Callable[..., Iterable[Any]]` makes a difference though, will look into it.

**Comment by rmitsch** [comment]
Ready for a final review round. We should decide whether we want to keep the file structure for tasks as-is - we could also merge `parser.py`, `example.py`, `scorer.py` into a single file?

File: spacy_llm/tasks/ner/task.py
Comment (svlandeg) [reply spacy_llm/tasks/ner/task.py]: Definitely not `examples`. I thought the current name was `prompt_examples`? Though I'm happy with `fewshot_examples`, too.

Comment (rmitsch) [reply spacy_llm/tasks/ner/task.py]: Some were `prompt_examples`, some were `examples`. I'll go with `fewshot_examples` then, I think it's the most recognizable term.

Comment (rmitsch) [reply spacy_llm/tasks/ner/task.py]: Done in https://github.com/explosion/spacy-llm/pull/242/commits/b443670e1a7c407fcb637e32b9088b8b9b5045b7. However, changing `examples` to `fewshot_examples` also in the factory functions (as it's currently the case in this PR) is a breaking change. Bumping task versions just for this also seems like a bother. 

Should we revert `fewshot_examples` to `examples` in the task factory functions?

Comment (svlandeg) [reply spacy_llm/tasks/ner/task.py]: As discussed internally, let's keep the registered task definitions as-is to avoid too many breaking changes in existing configs, and to minimize the refactoring done in this PR, which is difficult to track diff-wise.

**Comment by rmitsch** [comment]
`scorer` currently runs on `**kwargs` as well - we should update that to a generic protocol in a follow-up PR. Added an internal ticket.

File: spacy_llm/ty.py
Comment (kabirkhan) [reply spacy_llm/ty.py:84]: What about `ScorerProtocol` for the function implementation and `ScorableTaskProtocol` for if the task implements the scorer method?

Comment (rmitsch) [reply spacy_llm/ty.py]: Sounds good, changed in https://github.com/explosion/spacy-llm/pull/242/commits/aee012260cabe19625e7eb7f851e8cf9b2530ff4.

File: spacy_llm/tasks/spancat/registry.py
Comment (svlandeg) [inline spacy_llm/tasks/spancat/registry.py:32]:
```suggestion
```

Comment (svlandeg) [reply spacy_llm/tasks/spancat/registry.py:32]: (assuming that this wasn't an option to define in v1)

File: spacy_llm/tasks/spancat/task.py
Comment (svlandeg) [inline spacy_llm/tasks/spancat/task.py:35]: I think there's one inconsistency left here after all the renaming back-and-forth: `prompt_example_type` vs. `fewshot_example_type` (throughout the PR)

Comment (svlandeg) [reply spacy_llm/tasks/spancat/task.py:35]: Let's stick with `prompt_example_type` for now?

File: spacy_llm/tasks/summarization/parser.py
Comment (svlandeg) [inline spacy_llm/tasks/summarization/parser.py:15]:
```suggestion
```

File: spacy_llm/tasks/textcat/parser.py
Comment (svlandeg) [inline spacy_llm/tasks/textcat/parser.py:12]:
```suggestion
    """Parses LLM responses for spacy.TextCat.v1, v2 and v3
```

File: spacy_llm/ty.py
Comment (svlandeg) [inline spacy_llm/ty.py:72]: What is `ScorableCallableProtocol`?

File: spacy_llm/ty.py
Comment (svlandeg) [inline spacy_llm/ty.py:85]: ScorableTask?

File: spacy_llm/tasks/builtin_task.py
Comment (svlandeg) [inline spacy_llm/tasks/builtin_task.py:24]: ctrl-F and replace:
- LLMTaskProtocol
- TaskResponseParserProtocol

**Review by svlandeg — COMMENTED** [review]
This is looking good! Left a few final comments & fixes, nothing major.
Feel free to merge afterwards so we can move on to the CoT to put this in practice 🎉

**Comment by svlandeg** [comment]
And maybe to be sure, run the GPU & external tests at the end before merging.

File: spacy_llm/tasks/builtin_task.py
Comment (rmitsch) [reply spacy_llm/tasks/builtin_task.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/b687d54a5b3f72dfd05f848e490c140b45d70a3b.

File: spacy_llm/ty.py
Comment (rmitsch) [reply spacy_llm/ty.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/b687d54a5b3f72dfd05f848e490c140b45d70a3b.

File: spacy_llm/ty.py
Comment (rmitsch) [reply spacy_llm/ty.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/b687d54a5b3f72dfd05f848e490c140b45d70a3b.

File: spacy_llm/tasks/lemma/registry.py
Comment (rmitsch) [reply spacy_llm/tasks/lemma/registry.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/b687d54a5b3f72dfd05f848e490c140b45d70a3b.

File: spacy_llm/tasks/spancat/task.py
Comment (rmitsch) [reply spacy_llm/tasks/spancat/task.py]: Updated in https://github.com/explosion/spacy-llm/pull/242/commits/b804c5b48bf8b69f1271844ea77d88537a6bc07f.

File: spacy_llm/tasks/builtin_task.py
Comment (rmitsch) [reply spacy_llm/tasks/builtin_task.py]: Fixed in https://github.com/explosion/spacy-llm/pull/242/commits/b804c5b48bf8b69f1271844ea77d88537a6bc07f.

**Review by svlandeg — APPROVED** [review]
🎉

**Pull request merged**
